components/storage/indexeddb/mod.rs

```diff
@@ -194,6 +240,134 @@ impl<E: KvsEngine> IndexedDBEnvironment<E> {
+            }
+            if Self::is_readwrite_like(other.kvs.mode) && Self::scopes_overlap(entry, other) {
+                return false;
+            }
+            true
+        })
+    }
+
+    // https://w3c.github.io/IndexedDB/#transaction-scheduling
+    // A read/write transaction tx can start when there are no transactions which:
+    // Were created before tx; and
+    // have overlapping scopes with tx; and
+    // are not finished.
+    fn can_start_readwrite(&self, txn_id: u64) -> bool {
+        let Some(entry) = self.transactions.get(&txn_id) else {
+            return false;
+        };
+        self.transactions.iter().all(|(other_id, other)| {
+            if *other_id == txn_id {
+                return true;
+            }
+            if other.finished || other.creation_order >= entry.creation_order {
+                return true;
+            }
+            if Self::scopes_overlap(entry, other) {
+                return false;
+            }
+            true
+        })
+    }
+
+    // Scheduling happens in the backend so scope and ordering are evaluated against queued work.
+    // https://w3c.github.io/IndexedDB/#transaction-scheduling
+    fn schedule_transactions(
+        &mut self,
+        origin: ImmutableOrigin,
+        db_name: String,
+        sender: GenericSender<IndexedDBThreadMsg>,
+    ) {
+        // https://w3c.github.io/IndexedDB/#transaction-scheduling
+        // If multiple read/write transactions are attempting to access the same object store
+        // the transaction that was created first is the transaction which gets access, until the transaction is finished.
+        let mut ordered: Vec<(u64, u64)> = self
+            .transactions
+            .iter()
+            .filter(|(_, entry)| !entry.finished)
+            .map(|(id, entry)| (*id, entry.creation_order))
+            .collect();
+        ordered.sort_by_key(|(_, order)| *order);
+
+        for (txn_id, _) in ordered {
+            let should_start = {
+                let Some(entry) = self.transactions.get(&txn_id) else {
+                    continue;
+                };
+                if entry.started || entry.finished {
+                    continue;
+                }
+                match entry.kvs.mode {
+                    IndexedDBTxnMode::Readonly => self.can_start_readonly(txn_id),
+                    IndexedDBTxnMode::Readwrite | IndexedDBTxnMode::Versionchange => {
+                        self.can_start_readwrite(txn_id)
+                    },
+                }
+            };
+
+            if should_start {
+                if let Some(entry) = self.transactions.get_mut(&txn_id) {
+                    entry.started = true;
+                }
+                // https://w3c.github.io/IndexedDB/#transaction-lifecycle
+                // the implementation must queue a database task to start the transaction asynchronously.
+                let _ = sender.send(IndexedDBThreadMsg::StartQueuedTransaction {
```

Comment:
I don't understand why the spec says to queue a task, since this seems to be a message to the backend?